### PR TITLE
[mypyc] Remove unused labels and gotos

### PR DESCRIPTION
**Pull request by dosisod**
This PR removes labels which are never jumped to, as well as removing gotos that jump directly to the next label (and removes the now redundant label). I had a small issue with the `GetAttr` op code as it would insert C code after the end of the branch which could not be detected at the IR level, as seen in https://github.com/mypyc/mypyc/issues/600#issuecomment-1319530780 .

Here are some basic stats on the file sizes before and after this PR:

Building with `MYPY_OPT_LEVEL=0`, `MYPY_DEBUG_LEVEL=1`:

| Branch   | `.so` size      | `.c` size       | `.c` lines      |
|----------|-----------------|-----------------|-----------------|
| `master` | 43.9 MB         | 79.7 MB         | 2290675         |
| This PR  | 43.6 MB (-0.6%) | 78.6 MB (-1.4%) | 2179967 (-4.8%) |

Building with `MYPY_OPT_LEVEL=3`, `MYPY_DEBUG_LEVEL=0`:

| Branch   | `.so` size | `.c` size | `.c` lines |
|----------|------------|-----------|------------|
| `master` | 26.5 MB    | 79.7 MB   | 2291316    |
| This PR  | 26.5 MB    | 78.7 MB   | 2179701    |

(I don't know why the number of lines changes between optimization levels. It's only a couple hundred lines, perhaps Mypyc is emitting different code depending on optimization level?)

Unoptimized builds seem to benefit the most, while optimized builds look pretty much the same. I didn't check to see if they are *exactly* the same binary, but they are probably close if not the same. The biggest win is the drop in the number of lines C code being generated.

Closes https://github.com/mypyc/mypyc/issues/600

### CODE REVIEW

**Comment by hauntsaninja** [comment]
(Note the 32-bit test started failing on master and is unrelated to your PR)

**Review by ichard26 — CHANGES_REQUESTED** [review]
Hey thank you for the PR, your perseverance is appreciated :)

I'm not finished reviewing — I'd like to double check some more potentially problematic ops — but here are my initial comments. Overall, it looks good. It's impressive that this reduces the self-compile LOC by ~5%.

As usual, it's worth noting I'm a new reviewer (and not a committer!) so I may get some things wrong. Please speak up if I do!

File: mypyc/codegen/emitfunc.py
Comment (dosisod) [reply mypyc/codegen/emitfunc.py:240]: I just tried undoing this hunk, and Mypyc won't compile without it:

```
      build/__native_26051150be1624671934.c:7975:26: error: use of undeclared label 'CPyL11'
          if (!cpy_r_r28) goto CPyL11;
                               ^
      build/__native_26051150be1624671934.c:85644:26: error: use of undeclared label 'CPyL21'
          if (!cpy_r_r92) goto CPyL21;
                               ^
      build/__native_26051150be1624671934.c:343800:26: error: use of undeclared label 'CPyL18'
          if (!cpy_r_r56) goto CPyL18;
                               ^
      build/__native_26051150be1624671934.c:374890:26: error: use of undeclared label 'CPyL14'
          if (!cpy_r_r53) goto CPyL14;
                               ^
      build/__native_26051150be1624671934.c:411655:26: error: use of undeclared label 'CPyL18'
          if (!cpy_r_r63) goto CPyL18;
                               ^
      build/__native_26051150be1624671934.c:708426:26: error: use of undeclared label 'CPyL10'
          if (!cpy_r_r32) goto CPyL10;
                               ^
...
```

I assume that there are places in the code where a `Branch` is created/updated in such a way that the `true` and `false` branches are the same. I agree that this shouldn't be possible, but it is not enforced. Perhaps we should use an `assert` to ensure that the `true` and `false` branches cannot be the same?

File: mypyc/codegen/emitfunc.py
Comment (dosisod) [reply mypyc/codegen/emitfunc.py]: Yeah, I was using `bb` for `BasicBlock` which wasn't very descriptive. Since we are iterating using `targets()` I renamed it to `target` instead.

File: mypyc/codegen/emitfunc.py
Comment (dosisod) [reply mypyc/codegen/emitfunc.py:243]: I added an `assert` just to check, it seems to be getting stuck on `if TYPE_CHECKING` branches (might just be a coincidence).

Debug/asserts I added:

```diff
     def visit_branch(self, op: Branch) -> None:
+        true, false = op.true, op.false
+
+        print(op.value, op.rare)
+        print(self.module_name, self.source_path, op.line)
+
+        assert true is not false
```

Output:

```
...

      <mypyc.ir.ops.CallC object at 0x7f7b49e684c0> False
      mypy.argmap mypy/argmap.py 20
      <mypyc.ir.ops.Unbox object at 0x7f7b49e684f0> False
      mypy.argmap mypy/argmap.py 20
      <mypyc.ir.ops.Unbox object at 0x7f7b49e684f0> False
      mypy.argmap mypy/argmap.py -1
      Traceback (most recent call last):
        File "mypy/.venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "mypy/.venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "mypy/.venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 144, in get_requires_for_build_editable
          return hook(config_settings)
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 450, in get_requires_for_build_editable
          return self.get_requires_for_build_wheel(config_settings)
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 323, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 338, in run_setup
          exec(code, locals())
        File "<string>", line 169, in <module>
        File "mypy/mypyc/build.py", line 511, in mypycify
          groups, group_cfilenames = mypyc_build(
        File "mypy/mypyc/build.py", line 421, in mypyc_build
          group_cfiles, ops_text = generate_c(
        File "mypy/mypyc/build.py", line 229, in generate_c
          modules, ctext = emitmodule.compile_modules_to_c(
        File "mypy/mypyc/codegen/emitmodule.py", line 426, in compile_modules_to_c
          ctext = compile_ir_to_c(groups, modules, result, mapper, compiler_options)
        File "mypy/mypyc/codegen/emitmodule.py", line 310, in compile_ir_to_c
          ctext[group_name] = generator.generate_c_for_modules()
        File "mypy/mypyc/codegen/emitmodule.py", line 550, in generate_c_for_modules
          generate_native_function(fn, emitter, self.source_paths[module_name], module_name)
        File "mypy/mypyc/codegen/emitfunc.py", line 164, in generate_native_function
          ops[visitor.op_index].accept(visitor)
        File "mypy/mypyc/ir/ops.py", line 410, in accept
          return visitor.visit_branch(self)
        File "mypy/mypyc/codegen/emitfunc.py", line 204, in visit_branch
          assert true is not false
      AssertionError
```

And this is what is on line 20 of `mypy/argmap.py`:

https://github.com/python/mypy/blob/25047b87ffb961b787481af39ce428ef52fe08bd/mypy/argmap.py#L20-L21

Comment (ichard26) [reply mypyc/codegen/emitfunc.py:243]: Hmm yeah. I took a look into this and it's nontrivial to fix. This is happening because those imports are marked as internal and aren't processed, so the if statement's body block contains only a Goto op to the next block. Then, an optimization pass in the dataflow analysis deletes the body block, updating the branch op to directly jump to the next block (and consequently setting `true` and `false` to same block).

Let's leave this for a separate issue. Thanks for digging into this!

File: mypyc/codegen/emitfunc.py
Comment (ichard26) [inline mypyc/codegen/emitfunc.py:129]: Only labels are eliminated here, no?

**Review by ichard26 — APPROVED** [review]
LGTM*. I wasn't able to break your PR with the Cast op :)

Although... Cast ops still can merge two blocks (just like GetAttr) so it could have a possibility of breaking the control flow (as the branch gotos might not be at the end). I couldn't get it to break as a failed cast seemingly never jumps to the next block (unlike GetAttr). Honestly, I'm still pretty hesitant with the correctness / robustness of this PR.

Anyway, let's wait for Jukka's review now. Thanks!

File: mypyc/codegen/emitfunc.py
Comment (dosisod) [reply mypyc/codegen/emitfunc.py:129]: This code only marks the basic blocks as "referenced" if they meet the conditions above, nothing actually gets eliminated here. And yes, during codegen, labels won't be emitted for these "unreferenced" blocks, same for `goto`s (assuming they are jumping to the next block).

File: mypyc/codegen/emitfunc.py
Comment (mr-c) [reply mypyc/codegen/emitfunc.py:243]: @ichard26 Was an issue filed? I'm getting a `get_requires_for_build_editable` failure with mypy-1.4.0

https://github.com/common-workflow-language/cwltool/actions/runs/5353323117/jobs/9730454728#step:8:114

**Pull request merged**
